feat: Type inference in 'have been called with' parameters

eyalroth opened the pull request:
Fixes #15034.

## Summary

See #15034.

This feature has been attempted at the past (#13268), but introduced a breaking change (#13337) with overloaded functions, and thus was reverted.

To reach the desired typing -- especially the handling of overloaded functions vs no-args function vs the default `jest.fn()` type -- I eventually found [this code](https://github.com/microsoft/TypeScript/issues/32164#issuecomment-1072529475) and modified it slightly to fit better to the use-case.

## Test plan

See the tests added to the PR.

mrazauskas commented:
Interesting to see some progress in this field. Some time ago I was walking more or less the same path. All was fun until I realised that asymmetric matcher can be also nested. Also that expected value and target arguments even can have different shapes:

```ts
expect(
  jestExpect(
    jest.fn<(date: Date, name: [string, string]) => void>(),
  ).toHaveBeenCalledWith(jestExpect.any(Date), [
    jestExpect.any(String),
    jestExpect.any(String),
  ]),
).type.toBeVoid();

expect(
  jestExpect(
    jest.fn<(date: Date, name: [string, string]) => void>(),
  ).toHaveBeenCalledWith(jestExpect.any(Date), jestExpect.any(Array)),
).type.toBeVoid();
```

eyalroth commented:
> asymmetric matcher can be also nested

@mrazauskas Thanks, this is a good use/test case!

I think it's solvable with changing the current `WithAsymmetricMatchers` from:
```ts
type WithAsymmetricMatchers<P extends Array<any>> =
  Array<unknown> extends P ? never : {[K in keyof P]: P[K] | AsymmetricMatcher};
```
to:
```ts
type WithAsymmetricMatchers<P extends Array<any>> =
  Array<unknown> extends P
    ? never
    : {[K in keyof P]: DeepAsymmetricMatcher<P[K]>};

type DeepAsymmetricMatcher<T> = T extends object
  ? AsymmetricMatcher | {[K in keyof T]: DeepAsymmetricMatcher<T[K]>}
  : AsymmetricMatcher | T;
```

I would've preferred that all the asymmetric-matchers-producing functions will return the type that the matcher is supposed to stand for:
```ts
jestExpect.stringContaining('') // returns `string`
jestExpect.any(Number) // returns `number`
```
As it will allow for even better type checking. However, I'm not sure it's possible, and even if it is, it's likely going to break a lot of things.

---

Do you mind addressing to questions regarding this PR that I [posted in the issue](https://github.com/jestjs/jest/issues/15034#issuecomment-2176233017)?

> Before completing it, I wanted to know if that's the right direction.
> 
> I also had two specific questions about the PR:
> 
> 1. I added an import type from jest-mock to jest-types package.
> Should I change something in the dependencies?
> 
> 2. The type-checking of "examples and tests" (yarn typecheck:tests) fails on a few cases:
> a) In throwMatcher.test.ts, mockedMatch is defined with jest.fn and a no-args function, causing later expect to fail since it passes an argument to toHaveBeenCalledWith.
> b) In Farm.test.ts the same is happening with computeWorkerKey.
> c) In BaseWorkerPool.test.ts the type of the argument given to expect(Worker).toHaveBeenNthCalledWith() does not exactly match the type of the mock argument (WorkerOptions).
> 
> These failures seems to me like the expected behavior of this new feature, meaning that it's likely to introduce breaking changes to existing code. What are your thoughts on this?

mrazauskas commented:
1. You should add `jest-mock` to `devDependencies` of ` jest-types` and list it in `references` of `packages/jest-types/tsconfig.json`.
2. As you noticed, the failures are expected. Since Jest currently is releasing alphas, it is fine to land any breaking changes. Just update the tests to make the check pass.

By the way, I updated the example in my previous comment. Finally I recalled that the biggest challenge was to handle optional arguments like `name?`.

eyalroth commented:
@mrazauskas

> 1. You should add `jest-mock` to `devDependencies` of ` jest-types` and list it in `references` of `packages/jest-types/tsconfig.json`.

That didn't work for me. `yarn build` failed on compilation `tsc`.

> By the way, I updated the example in my previous comment. Finally I recalled that the biggest challenge was to handle optional arguments like `name?`.

The code seems to handle this as well. See my updated commit with all the new test cases.

mrazauskas commented:
Ah.. Both `package.json` and `tsconfig.json` in `jest-mock` are listing `@jest/types` as a dependency. That is a mistake, because the dependency was removed in #7971 (five years ago). This can be cleaned up (in separate PR).

I see currently you are importing one type from `jest-mock` to `expect`. Is it possible to declare `jest-mock` as a devDependency in the `expect` package? Or build script is complaining?

Some time later I will try out updated types. Really interesting.

eyalroth commented:
> I see currently you are importing one type from `jest-mock` to `expect`. Is it possible to declare `jest-mock` as a devDependency in the `expect` package? Or build script is complaining?

@mrazauskas That works 😃 I amended my last commit with this addtional change.

mrazauskas commented:
If I got it right, your implementation only supports type inference for the `Mock` type. Only wondering why did you decide so? I mean, args of in the following (this is one examples from Jest's docs) is still inferred as `Array<unknown>`: 


```ts
jest.spyOn(globalThis, 'setTimeout');

jest(setTimeout).toHaveBeenLastCalledWith(jest.any(Function), 1000);
```

Not a blocker. That is fine, of course. I just wanted to check if that is by design or this got overlooked?

eyalroth commented:
> your implementation only supports type inference for the `Mock` type

@mrazauskas That's an oversight :)

I adjusted the code to support both `MockInstance` (broader than `Mock`) and functions without mock type wrapping. I also added a few minimal test cases for these mock-less functions.

SimenB commented on packages/expect/package.json:
```diff
@@ -31,7 +31,8 @@
     "chalk": "^4.0.0",
-    "immutable": "^4.0.0"
+    "immutable": "^4.0.0",
+    "jest-mock": "workspace:*"
```
should be in `dependencies`

SimenB commented on packages/jest-types/__typetests__/expect.test.ts:
I love that you've added a bunch of tests (thanks!), but this is getting pretty big. maybe we should have a `packages/jest-types/__typetests__/expect/mocks.test.ts` or something? or `packages/jest-types/__typetests__/expect/toHaveBeenCalled.test.ts` to group by matcher?

eyalroth replied:
@SimenB My thoughts exactly :)

I like the idea of having an `expect/` directory with a file per function. I'll see about arranging the PR so it's clear which tests were moved and which were added.

SimenB commented on packages/expect/src/types.ts:
```diff
@@ -307,3 +308,148 @@ export interface Matchers<R extends void | Promise<void>, T = unknown> {
   toThrow(expected?: unknown): R;
 }
+
+/**
```
thanks for the extensive comments 👍

SimenB commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@
 - `[jest-environment-jsdom]` [**BREAKING**] Upgrade JSDOM to v22 ([#13825](https://github.com/jestjs/jest/pull/13825))
 - `[@jest/environment-jsdom-abstract]` Introduce new package which abstracts over the `jsdom` environment, allowing usage of custom versions of JSDOM ([#14717](https://github.com/jestjs/jest/pull/14717))
 - `[jest-environment-node]` Update jest environment with dispose symbols `Symbol` ([#14888](https://github.com/jestjs/jest/pull/14888) & [#14909](https://github.com/jestjs/jest/pull/14909))
+- `[expect, @jest/expect]` [**BREAKING**] Add type inference for function parameters in `CalledWith` assertions ([#15129](https://github.com/facebook/jest/pull/15129))
```
why is this a breaking change? the odl type was just `Function` or something?

also, only `expect` seems to be impacted? I don't see any changes to  `@jest/expect` at least 😀

eyalroth replied:
@SimenB I'm thinking it will likely break some existing tests that use "have been called" functions due to incorrect typing, like it did in tests/examples in this repo. I'm not sure what are the full implications of this "tag", so obviously I will let you decide.

As for `@jest/expect`, I merely copied it from the last PR that tried to introduce this feature. I can obviously remove it.

SimenB replied:
It's just used as a way to catch the reader's eye.

I really should have included a migration guide thing when I started landing breaking changes, where _how_ the breaking change might affect you would be listed 😅 I'll have to add one

SimenB reviewed (approved):
thanks for working on this!

and thanks for reviewing @mrazauskas 🙏

eyalroth commented:
@mrazauskas @SimenB Thanks for your help on this PR. I couldn't have done it without you ❤️ 

Please let me know if you have more concerns.

As for the changelog, I'm keeping it for you to decide whether to include the "breaking" tag and whether `@jest/expect` should be mentioned as well.

mrazauskas commented:
@eyalroth I was playing with updated types. All works as expected. Great job!
